src/Psalm/Internal/Analyzer/Statements/ExpressionAnalyzer.php

```diff
@@ -147,12 +147,27 @@ public static function analyze(
                     $stmt->inferredType = Type::getLiteralClassString($context->self);
                     break;
 
+                case '__namespace__':
+                    $namespace = $statements_analyzer->getNamespace();
+                    if ($namespace === null &&
+                    IssueBuffer::accepts(
+                        new UndefinedConstant(
+                            'Cannot get __namespace__ outside a namespace',
+                            new CodeLocation($statements_analyzer->getSource(), $stmt)
+                        ),
+                        $statements_analyzer->getSuppressedIssues()
+                    )) {
+                        // fall through
+                    }
+
+                    $stmt->inferredType = Type::getString($namespace);
+                    break;
+
                 case '__file__':
```

Comment:
Aside, out of the scope of this PR: All of this code was probably written before Psalm added support for literal string/int types.

It's useful to have the literal values for these constants:

- __FUNCTION__ for when validating that a callback is a valid callback
- __DIR__/__FILE__ for various checks, e.g. checking that calls to require_once with complex expressions are valid PHP files (absolute path is the most useful)
- __TRAIT__ for the same reason as __CLASS__, for checking for valid class-likes (make sure to remove leading the `\\` character)

Reply:
I agree, and yes - best left out of scope for this PR.